src/main/java/com/ibm/watson/modelmesh/payload/AsyncPayloadProcessor.java

```diff
@@ -0,0 +1,124 @@
+ * of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.ibm.watson.modelmesh.payload;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
+
+public class AsyncPayloadProcessor implements PayloadProcessor {
+
+    private static final Logger logger = LoggerFactory.getLogger(AsyncPayloadProcessor.class);
+
+    private final PayloadProcessor delegate;
+
+    private final FixedSizeConcurrentLinkedDeque<Payload> requestPayloads;
+    private final FixedSizeConcurrentLinkedDeque<Payload> responsePayloads;
+
+    public AsyncPayloadProcessor(PayloadProcessor delegate) {
+        this(delegate, 1, TimeUnit.MINUTES, Executors.newScheduledThreadPool(1), 1000);
+    }
+
+    public AsyncPayloadProcessor(PayloadProcessor delegate, int delay, TimeUnit timeUnit,
+                                 ScheduledExecutorService executorService, int capacity) {
+        this.delegate = delegate;
+        this.requestPayloads = new FixedSizeConcurrentLinkedDeque<>(capacity);
+        this.responsePayloads = new FixedSizeConcurrentLinkedDeque<>(capacity);
+
+        executorService.scheduleWithFixedDelay(() -> {
+            Payload p;
+            while ((p = requestPayloads.poll()) != null) {
+                delegate.processRequest(p);
+                maybeRelease(p);
+            }
+            int droppedRequest = requestPayloads.dropped.getAndSet(0);
+            if (droppedRequest > 0) {
+                logger.warn("{} request payloads were skipped because of {} capacity limit", droppedRequest, capacity);
+            }
+            while ((p = responsePayloads.poll()) != null) {
+                delegate.processResponse(p);
+                maybeRelease(p);
+            }
+            int droppedResponse = responsePayloads.dropped.getAndSet(0);
+            if (droppedResponse > 0) {
+                logger.warn("{} response payloads were skipped because of {} capacity limit", droppedResponse, capacity);
+            }
+        }, 0, delay, timeUnit);
+    }
+
+    private static void maybeRelease(Payload p) {
+        if (p.getData() != null) {
+            p.getData().release();
+        }
+    }
+
+    @Override
+    public String getName() {
+        return delegate.getName() + "-async";
+    }
+
+    @Override
+    public void processRequest(Payload payload) {
+        requestPayloads.offer(payload);
+        maybeRetain(payload);
```

Comment:
IIUTC, given the fact that the async processor will go through the queue for actual processing at fixed delays (e.g., 1m), issuing a `release` call right after the payload gets added to the queue would allow garbage collector to drop it even before the actual processing happens. 
That's why there'a `retain` call when a payload is added and a `release` call when the payload is actually processed (e.g., 1m after) in the scheduled thread. 
However you're right that we should only call `retain` if the payload gets into the queue, that is when `offer() == true`.

Reply:
Per my other comment I don't think there's any reason to have delays between processing the queue. But even if there is a delay you can't be sure that a payload added to the queue won't be instantly processed and released. Attempting to retain once the ref count has already reached 0 will cause an error to be thrown.

I still think the best though would be to return an boolean indicating whether or not the payload was successfully enqueued, and just have the outside release depend on that. Then no new retain is necessary anywhere.

Reply:
ok, implemented this in https://github.com/kserve/modelmesh/pull/84/commits/1ab932dbbdf09e9f86b32c926865426c7972daf0